src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -1777,6 +1779,39 @@ internal void EnsureNativeIntegerAttributeExists()
             EnsureEmbeddableAttributeExists(EmbeddableAttributes.NativeIntegerAttribute);
         }
 
+#nullable enable
+        /// <summary>
+        /// Creates the ThrowIfNull and Throw helpers if needed.
+        /// </summary>
+        /// <remarks>
+        /// The ThrowIfNull and Throw helpers are modeled off of the helpers on ArgumentNullException.
+        /// https://github.com/dotnet/runtime/blob/22663769611ba89cd92d14cfcb76e287f8af2335/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L56-L69
+        /// </remarks>
+        internal MethodSymbol EnsureThrowIfNullFunctionExists(SyntaxNode syntaxNode, SyntheticBoundNodeFactory factory, DiagnosticBag? diagnostics)
+        {
+            // If we have already generated the helper, possibly for another switch
+            // or on another thread, we don't need to regenerate it.
+            var privateImplClass = GetPrivateImplClass(syntaxNode, diagnostics);
+            if (privateImplClass.GetMethod(CodeAnalysis.CodeGen.PrivateImplementationDetails.SynthesizedThrowIfNullFunctionName) is { } throwIfNullAdapter)
+            {
+                Debug.Assert(privateImplClass.GetMethod(CodeAnalysis.CodeGen.PrivateImplementationDetails.SynthesizedThrowFunctionName) != null);
+                return (MethodSymbol)throwIfNullAdapter.GetInternalSymbol()!;
+            }
+
+            TypeSymbol returnType = factory.SpecialType(SpecialType.System_Void);
+            TypeSymbol argumentType = factory.SpecialType(SpecialType.System_Object);
+            TypeSymbol paramNameType = factory.SpecialType(SpecialType.System_String);
+
+            var sourceModule = SourceModule;
+            var throwMethod = new SynthesizedThrowMethod(sourceModule, privateImplClass, returnType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwMethod.GetCciAdapter());
+
+            var throwIfNullMethod = new SynthesizedThrowIfNullMethod(sourceModule, privateImplClass, throwMethod, returnType, argumentType, paramNameType);
+            privateImplClass.TryAddSynthesizedMethod(throwIfNullMethod.GetCciAdapter());
```

Comment:
It feels like, if this returns false, we should be getting the final throwIfNull method that some other thread succeeded in adding, rather than returning the symbol we failed to add. It doesn't look like any use of `TryAddSynthesizedMethod` today is used by a method that returns a value, so this would be a new change.

Reply:
Ahh, you're right. It feels like if we fail to add the first synthesized method here, we need to spin-wait or something until we get the second synthesized method that we expect another thread to add.

Reply:
I wouldn't bother spin waiting on the first one, we'll do a bit of extra work, but not much. Instead just gate it on the second add, and re-do the lookup if it fails. (You may want to dig into how much work creating them actually is compared with the lookup and just always try and add them and only return the already created one if it fails)

Reply:
I went ahead and made this change, although after thinking about it a bit I suspect that the specific instance of the method symbol between the PrivateImplementationDetails class and a particular call site of the helper method is not going to end up affecting compiler behavior.

Reply:
It likely won't, until it ends up being exposed via some public API and then all hell will break loose.

Reply:
Fixed by using a lock.

Reply:
Well, fixed by adding-then-getting, accomplishing the same thing.